File: api/app/models/issue.rb
# frozen_string_literal: true

class Issue < ApplicationRecord
  validates :status,  presence: true
  validates :problem, presence: true, uniqueness: { scope: :team_id }
  validates :team,    presence: true

  # 状態遷移条件
  # unsolved: 未対応
  # in_progress: 対応中
  # solved: 解決
  enum status: {
    unsolved: 1,
    in_progress: 2,
    solved: 3
  }

  belongs_to :problem
  belongs_to :team
  has_many :comments, dependent: :destroy, class_name: 'IssueComment'

  # staff以外には対応中を未対応と表示する
  def response_status(team:)
    return 'unsolved' if !team.staff? && in_progress?

    status
  end

  def transition_by_click(team:)
    case status
    when 'unsolved'
      self.status = team.staff? ? 'in_progress' : 'solved'
    when 'in_progress'
      self.status = 'solved'
    when 'solved'
      self.status = team.staff? ? 'in_progress' : 'unsolved'
    else
      raise UnhandledIssueStatus, status
    end
  end

  def transition_by_comment(team:)
    case status
    when 'unsolved'
      self.status = 'in_progress' if team.staff?
    when 'in_progress'
      self.status = 'unsolved' if team.player?
    when 'solved'
      self.status = 'unsolved' if team.player?
    else
      raise UnhandledIssueStatus, status
    end
  end
end

File: ui/orm/Issue.js
import orm from '~/orm'
import BaseModel from '~/orm/BaseModel'

export default class Issue extends BaseModel {
  static entity = 'issues'
  // 多段でfetchEagerが使えないのでここで明記
  static eagerLoad = ['comments']

  static fields() {
    return {
      id: this.string(),
      status: this.string(),
      comments: this.hasMany(orm.IssueComment, 'issueId'),
      problemId: this.string(),
      problem: this.belongsTo(orm.Problem, 'problemId'),
      teamId: this.string(),
      team: this.belongsTo(orm.Team, 'teamId'),
      updatedAt: this.string()
    }
  }

  get statusNum() {
    switch (this.status) {
      case 'unsolved':
        return 0
      case 'in_progress':
        return 1
      case 'solved':
        return 2
      default:
        throw new Error(`unsupported status ${this.status}`)
    }
  }

  get statusJp() {
    switch (this.status) {
      case 'unsolved':
        return '未解決'
      case 'in_progress':
        return '対応中'
      case 'solved':
        return '解決済'
      default:
        throw new Error(`unsupported status ${this.status}`)
    }
  }

  get statusColor() {
    switch (this.status) {
      case 'unsolved':
        return 'error'
      case 'in_progress':
        return 'warning'
      case 'solved':
        return 'success'
      default:
        throw new Error(`unsupported status ${this.status}`)
    }
  }

  get ourComments() {
    return this.comments.filter(c => c.isOurComment)
  }

  get theirsComments() {
    return this.comments.filter(c => !c.isOurComment)
  }

  get latestReplyAt() {
    const comment = $nuxt.findNewer(this.theirsComments)
    return comment ? comment.createdAt : null
  }

  get latestReplyAtDisplay() {
    const comment = $nuxt.findNewer(this.theirsComments)
    return comment ? comment.createdAtShort : 'なし'
  }
}
